diff --git a/packages/plugin-ext/src/hosted/node/hosted-instance-manager.ts b/packages/plugin-ext/src/hosted/node/hosted-instance-manager.ts
index b7352767..688898c6 100644
--- a/packages/plugin-ext/src/hosted/node/hosted-instance-manager.ts
+++ b/packages/plugin-ext/src/hosted/node/hosted-instance-manager.ts
@@ -26,6 +26,7 @@ import { LogType } from './../../common/types';
 import { HostedPluginUriPostProcessor, HostedPluginUriPostProcessorSymbolName } from './hosted-plugin-uri-postprocessor';
 import { HostedPluginSupport } from './hosted-plugin';
 import { DebugConfiguration } from '../../common';
+import { isElectron } from '@theia/core';
 const processTree = require('ps-tree');
 
 export const HostedInstanceManager = Symbol('HostedInstanceManager');
@@ -232,7 +233,21 @@ export abstract class AbstractHostedInstanceManager implements HostedInstanceMan
     }
 
     protected async getStartCommand(port?: number, debugConfig?: DebugConfiguration): Promise<string[]> {
-        const command = ['yarn', 'theia', 'start'];
+
+        const processArguments = process.argv;
+        let command: string[];
+        if (isElectron()) {
+            command = ['yarn', 'theia', 'start'];
+        } else {
+            command = processArguments.filter(arg => {
+                // remove --port= argument if set
+                if (arg.startsWith('--port=')) {
+                    return;
+                } else {
+                    return arg;
+                }
+            });
+        }
         if (process.env.HOSTED_PLUGIN_HOSTNAME) {
             command.push('--hostname=' + process.env.HOSTED_PLUGIN_HOSTNAME);
         }
