--- src/app/index.module.ts
+++ src/app/index.module.ts
@@ -32,12 +32,36 @@
 import {StacksConfig} from './stacks/stacks-config';
 import {DemoComponentsCtrl} from './demo-components/demo-components.controller';
-
+import * as Keycloak from '../components/keycloak/keycloak';
+
+
+let keycloakConfig = {
+  "url": "http://172.17.0.1:5050/auth",
+  "realm": "che",
+  "clientId": "che-public",
+};

 // init module
 let initModule = angular.module('userDashboard', ['ngAnimate', 'ngCookies', 'ngTouch', 'ngSanitize', 'ngResource', 'ngRoute',
   'angular-websocket', 'ui.bootstrap', 'ui.codemirror', 'ngMaterial', 'ngMessages', 'angularMoment', 'angular.filter',
   'ngDropdowns', 'ngLodash', 'angularCharts', 'ngClipboard', 'uuid4', 'angularFileUpload']);

+angular.element(document).ready(() => {
+  console.log('running keycloak init sequence');
+
+  window['_keycloak'] = Keycloak(keycloakConfig);
+
+  window['_keycloak']
+    .init({
+      onLoad: 'login-required'
+    })
+    .success(() => {
+      angular.bootstrap(document, ['userDashboard'], {strictDi:true}); // manually bootstrap Angular
+    });
+});
+
+initModule.factory('keycloak', $window => {
+  return $window._keycloak;
+});

 // add a global resolve flag on all routes (user needs to be resolved first)
 initModule.config(['$routeProvider', ($routeProvider) => {
@@ -54,7 +78,6 @@
           deferred.reject(error);
         });
       }
-
       return deferred.promise;
     }];

@@ -65,6 +88,7 @@
     route.resolve || (route.resolve = {});
     route.resolve.app = ['$q', 'chePreferences', ($q, chePreferences) => {
       let deferred = $q.defer();
+
       if (chePreferences.getPreferences()) {
         deferred.resolve();
       } else {
@@ -74,9 +98,9 @@
           deferred.reject(error);
         });
       }
-
       return deferred.promise;
     }];
+
     return $routeProvider.otherwise(route);
   };

@@ -110,12 +134,11 @@
 /**
  * Setup route redirect module
  */
-initModule.run(['$rootScope', '$location', '$routeParams', 'routingRedirect', '$timeout', 'ideIFrameSvc', 'cheIdeFetcher', 'routeHistory', 'cheUIElementsInjectorService', 'workspaceDetailsService',
+initModule.run(['$rootScope', '$location', '$routeParams', 'routingRedirect', '$timeout','ideIFrameSvc', 'cheIdeFetcher', 'routeHistory', 'cheUIElementsInjectorService', 'workspaceDetailsService',
   ($rootScope, $location, $routeParams, routingRedirect, $timeout, ideIFrameSvc, cheIdeFetcher, routeHistory, cheUIElementsInjectorService, workspaceDetailsService) => {
     $rootScope.hideLoader = false;
     $rootScope.waitingLoaded = false;
     $rootScope.showIDE = false;
-
     workspaceDetailsService.addPage('Projects', '<workspace-details-projects></workspace-details-projects>', 'icon-ic_inbox_24px');
     workspaceDetailsService.addPage('SSH', '<workspace-details-ssh></workspace-details-ssh>', 'icon-ic_vpn_key_24px');

@@ -166,6 +189,38 @@
   }]);


+initModule.factory('KeyCloakTokenInterceptor', ($q, keycloak) => {
+  return {
+    request: (config) => {
+      if (config.url.indexOf(keycloakConfig.url) > -1) {
+        return config;
+      }
+      if (keycloak && keycloak.token) {
+        let deferred = $q.defer();
+        keycloak.updateToken(5).success(function () {
+          config.headers = config.headers || {};
+          config.headers.Authorization = 'Bearer ' + keycloak.token;
+          deferred.resolve(config);
+        }).error(function () {
+          console.log('token refresh failed :' + config.url);
+          deferred.reject('Failed to refresh token');
+        });
+        return deferred.promise;
+      }
+      return config || $q.when(config);
+    },
+    response: (response) => {
+      return response || $q.when(response);
+    },responseError: (rejection)=>{
+        return $q.reject(rejection);
+    },
+    requestError: (rejection) =>{
+       return $q.reject(rejection);
+    }
+
+  }
+});
+
 // add interceptors
 initModule.factory('ETagInterceptor', ($window, $cookies, $q) => {

@@ -355,8 +415,10 @@
 });

 initModule.config(['$routeProvider', '$locationProvider', '$httpProvider', ($routeProvider, $locationProvider, $httpProvider) => {
+
+ $httpProvider.interceptors.push('KeyCloakTokenInterceptor');
   // add the ETag interceptor for Che API
-  $httpProvider.interceptors.push('ETagInterceptor');
+ $httpProvider.interceptors.push('ETagInterceptor');
 }]);


