--- gulp/scripts.js
+++ gulp/scripts.js
@@ -26,7 +26,7 @@
     watch: watch,
     module: {
       preLoaders: [{ test: /\.ts$/, exclude: /node_modules/, loader: 'tslint-loader'}],
-      loaders: [{ test: /\.ts$/, exclude: /node_modules/, loaders: ['ng-annotate', 'babel-loader', 'awesome-typescript-loader']}]
+      loaders: [{ test: /\.(ts)$/, exclude: /node_modules/, loaders: ['ng-annotate', 'babel-loader', 'awesome-typescript-loader']}]
     },
     output: { filename: 'index.module.js' }
   };

